cloudevents/generic/baseclient.go

```diff
@@ -0,0 +1,140 @@
+package generic
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+
+	"open-cluster-management.io/api/cloudevents/generic/options"
+)
+
+type receiveFn func(ctx context.Context, evt cloudevents.Event)
+
+type baseClient struct {
+	sync.RWMutex
+	cloudEventsOptions     options.CloudEventsOptions
+	cloudEventsClient      cloudevents.Client
+	cloudEventsErrChan     <-chan error
+	cloudEventsRateLimiter flowcontrol.RateLimiter
+	subStarted             bool
+}
+
+func (c *baseClient) publish(ctx context.Context, evt cloudevents.Event) error {
+	now := time.Now()
+
+	if err := c.cloudEventsRateLimiter.Wait(ctx); err != nil {
+		return fmt.Errorf("client rate limiter Wait returned an error: %w", err)
+	}
+
+	latency := time.Since(now)
+	if latency > longThrottleLatency {
+		klog.Warningf(fmt.Sprintf("Waited for %v due to client-side throttling, not priority and fairness, request: %s",
+			latency, evt))
+	}
+
+	sendingCtx, err := c.cloudEventsOptions.WithContext(ctx, evt.Context)
+	if err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Sent event: %v\n%s", ctx, evt)
+
+	if result := c.cloudEventsClient.Send(sendingCtx, evt); cloudevents.IsUndelivered(result) {
+		return fmt.Errorf("failed to send event %s, %v", evt, result)
+	}
+
+	return nil
+}
+
+func (c *baseClient) subscribe(ctx context.Context, receive receiveFn) {
+	c.Lock()
+	defer c.Unlock()
+
+	// make sure there is only one subscription go routine starting for one client.
+	if c.subStarted {
+		klog.Warningf("the subscription has already started")
+		return
+	}
+
+	// start a go routine to handle cloudevents subscription
+	go func() {
+		var receiverCtx context.Context
+		var receiverCancel context.CancelFunc
+		var retrying, receiving bool
+
+		// the reconnect backoff will stop at [1,5) min interval. If we don't backoff for 10min, we reset the backoff.
+		connBackoffManager := wait.NewExponentialBackoffManager(5*time.Second, 1*time.Minute, 10*time.Minute, 5.0, 1.0, &clock.RealClock{})
+
+		for {
+			select {
+			case <-ctx.Done():
+				if receiverCancel != nil {
+					receiverCancel()
+				}
+				return
+			case err, ok := <-c.cloudEventsErrChan:
+				if !ok {
+					continue
+				}
+
+				klog.Errorf("the cloudevents client is disconnected, %v", err)
+
+				// the cloudevents network connection is closed, stop the receiver and start retry
+				if receiverCancel != nil {
+					receiverCancel()
+				}
+				retrying = true
```

Comment:
the cloudEventsClient is also used by send method, if we set it to `nil` here, the send may meet the nil pointer, so I did not set it to nil, what's your suggestion?

Reply:
even though, the send will fail right? I think there might need a rwlock in send otherwise there is always a change that the two threads race.

Reply:
I mean after backoff it will go to select in the next loop while client is not instantiated, and it will not have a chance to receive message from c.cloudEventsErrChan

Reply:
yes, the send will be fail, we maybe need a lock here for the client

the c.cloudEventsErrChan will be only send by a running client, so if the client is retrying, the select will not be blocked the next retry